Implement get all installed extensions

sannimichaelse opened the pull request:
Show All Installed Extensions

Currently, the Service mesh details table shows some details but would be great to add installed extensions to it

Add a new section called `Installed Extensions`, Modify get extension endpoint to List all installed extensions as well support getting single ones

Once merged, The installed extensions section should be list all installed extensions

Fixes #5899

Signed-off-by: Sanni Michael <sannimichaelse@gmail.com>

Previous Look

![image](https://user-images.githubusercontent.com/37812752/123558076-63853680-d78c-11eb-9fd7-c6830dc32317.png)

New Look

![image](https://user-images.githubusercontent.com/37812752/123558109-95969880-d78c-11eb-89bb-1eff847a3ac7.png)



<!--  Thanks for sending a pull request!

If you already have a well-structured git commit message, chances are GitHub
set the title and description of this PR to the git commit message subject and
body, respectively. If so, you may delete these instructions and submit your PR.

If this is your first time, please read our contributor guide:
https://github.com/linkerd/linkerd2/blob/main/CONTRIBUTING.md

The title and description of your Pull Request should match the git commit
subject and body, respectively. Git commit messages are structured as follows:

```
Subject

Problem

Solution

Validation

Fixes #[GitHub issue ID]

DCO Sign off
```

Example git commit message:

```
Introduce Pull Request Template

GitHub's community guidelines recommend a pull request template, the repo was
lacking one.

Introduce a `PULL_REQUEST_TEMPLATE.md` file.

Once merged, the
[Community profile checklist](https://github.com/linkerd/linkerd2/community)
should indicate the repo now provides a pull request template.

Fixes #3321

Signed-off-by: Jane Smith <jane.smith@example.com>
```

Note the git commit message subject becomes the pull request title.

For more details around git commits, see the section on Committing in our
contributor guide:
https://github.com/linkerd/linkerd2/blob/main/CONTRIBUTING.md#committing
-->

Pothulapati commented:
Hey @sannimichaelse 👋🏼 

So, I just tried testing out this PR but couldn't actually see the extensions being displayed.

![image](https://user-images.githubusercontent.com/7892868/123593114-c79b0f80-d80b-11eb-9b8c-8fb24468db80.png)
As you can see, I have both `linkerd-viz` and `linkerd-smi` installed but still can't see them in the list. I can take a look later today on what's going wrong, if you are not able to find the issue. :)

sannimichaelse commented:
Hey @Pothulapati thanks for checking it out. Found the issue and fixed it already. I also squashed all the commits to keep the commit history clean. You can try to test again

![image](https://user-images.githubusercontent.com/37812752/123621699-653e1100-d803-11eb-9926-86fb932ea817.png)

mateiidavid commented on web/app/js/components/ServiceMesh.jsx:
```diff
@@ -109,6 +118,15 @@ class ServiceMesh extends React.Component {
   }
 
+  getInstalledExtensions() {
+    const { extensions } = this.state;
+    let extensionList = [];
+    if (!_isEmpty(extensions.extensions)) {
+      extensionList = extensions.extensions.map(extension => extension.metadata);
+    }
+    return extensionList;
```
I think it'd be easier here to just return the resulting array from map.

We declare an empty array on L123, if the state is not empty (and we have a list of extensions) then we map through and change an extension to be its associated metadata. If the array is empty then mapping through it will still give an empty array, right?

```suggestion
    const { extensions } = this.state;
    return extensions.extensions.map(extension => extension.metadata);
```

also, any way we can avoid doing `extensions.extensions` and instead do just `extensions.map`?

sannimichaelse replied:
So if the thing is, the test passes a different value to the service mesh component on component did mount. I noticed it was passing 

```
{ metrics: [] }
```

So the `extensions.extensions` resulted in undefined. And that's why the test was failing initially. I had to put a check to ensure it exists before mapping it

sannimichaelse replied:
For the wrapped heading - would check it out again and see why its like that

mateiidavid commented on web/app/js/components/util/ApiHelpers.jsx:
```diff
@@ -147,7 +147,11 @@ const ApiHelpers = (pathPrefix, defaultMetricsWindow = '1m') => {
+    if (name) {
+      extensionPath += `?extension_name=${name}`;
+    }
+    return apiFetch(extensionPath);
```
I think the name might be a bit misleading here. It seems that when we don't add a query param to the `GET /api/extension` path we return all of the extensions. I'm wondering if this shouldn't be its own handler and its own call.

Perhaps we might also be able to rename the path so the get call goes to `/api/extensions/` in which case `/api/extensions?extension_name=viz` might also make sense. Let me know if this makes sense and what your thoughts around this are.

sannimichaelse replied:
Yea thought about this too but from the comment from the issue it was stated that the api/extension route should be modified to accommodate getting all and single extensions. You can check more info on that on the original issue #5899

dadjeibaah replied:
I agree with @mateiidavid on renaming the endpoint. I think now that we are adding the capability to return a list of extensions, it makes more sense to rename the endpoint to `/api/extensions` as well as the method name. On the other hand, I think it would be ok to use one single handler for both cases since the info we dealing with are similar.

mateiidavid commented on web/srv/api_handlers.go:
```diff
@@ -439,18 +439,26 @@ func (h *handler) handleGetExtension(w http.ResponseWriter, req *http.Request, _
-	} else if err != nil {
-		renderJSONError(w, err, http.StatusInternalServerError)
-		return
+	if extensionName == "" {
+		extensions, err := h.k8sAPI.GetAllNamespacesWithExtensionLabel(ctx)
+		if err != nil {
+			renderJSONError(w, err, http.StatusInternalServerError)
+			return
+		}
+		resp["extensions"] = extensions
+	} else {
+		ns, err := h.k8sAPI.GetNamespaceWithExtensionLabel(ctx, extensionName)
+		if err != nil && strings.HasPrefix(err.Error(), "could not find") {
+			renderJSON(w, resp)
+			return
+		} else if err != nil {
+			renderJSONError(w, err, http.StatusInternalServerError)
+			return
+		}
+		resp["extensionName"] = ns.GetLabels()[k8s.LinkerdExtensionLabel]
+		resp["namespace"] = ns.Name
```
wdyt about reducing the nesting here with an early return?

```suggestion
	if extensionName == "" {
		extensions, err := h.k8sAPI.GetAllNamespacesWithExtensionLabel(ctx)
		if err != nil {
			renderJSONError(w, err, http.StatusInternalServerError)
			return
		}
		
		resp["extensions"] = extensions
		renderJSON(w, resp)
		return 
	} 
	
	ns, err := h.k8sAPI.GetNamespaceWithExtensionLabel(ctx, extensionName)
	if err != nil && strings.HasPrefix(err.Error(), "could not find") {
		renderJSON(w, resp)
		return
	} else if err != nil {
		renderJSONError(w, err, http.StatusInternalServerError)
		return
	}
	resp["extensionName"] = ns.GetLabels()[k8s.LinkerdExtensionLabel]
	resp["namespace"] = ns.Name
	
	renderJSON(w, resp)
```

sannimichaelse replied:
Yea this looks good, would update the PR. 

Thanks @mateiidavid

mateiidavid reviewed (commented):
Thank you @sannimichaelse! I built & deployed your changes to my local k3d cluster and it seems the issue with extensions not being rendered has been fixed,  can see all of my installed extensions :) 

I left some general comments on the code, curious what you think.

Also, a thing that stood out to me visually (and I noticed this in @Pothulapati's screenshot too), the heading seems to be wrapped. Is that something that we want? 

![image](https://user-images.githubusercontent.com/18265670/123631181-5d379e80-d80e-11eb-8d99-645ea435001f.png)

dadjeibaah commented on go.sum:
```diff
@@ -237,6 +237,8 @@ github.com/edsrzf/mmap-go v1.0.0/go.mod h1:YO35OhQPt3KJa3ryjFM5Bs14WD66h8eGKpfaB
 github.com/elazarl/goproxy v0.0.0-20180725130230-947c36da3153/go.mod h1:/Zj4wYkgs4iZTTu3o/KG3Itv/qCCa8VVMlb3i9OVuzc=
 github.com/elazarl/goproxy v0.0.0-20190711103511-473e67f1d7d2 h1:aZtFdDNWY/yH86JPR2WX/PN63635VsE/f/nXNPAbYxY=
 github.com/elazarl/goproxy v0.0.0-20190711103511-473e67f1d7d2/go.mod h1:/Zj4wYkgs4iZTTu3o/KG3Itv/qCCa8VVMlb3i9OVuzc=
+github.com/elazarl/goproxy/ext v0.0.0-20210110162100-a92cc753f88e h1:CQn2/8fi3kmpT9BTiHEELgdxAOQNVZc9GoPA4qnQzrs=
```
Are these changes related to this PR? If not I would back this out.

dadjeibaah commented on web/srv/api_handlers.go:
```diff
@@ -439,18 +439,26 @@ func (h *handler) handleGetExtension(w http.ResponseWriter, req *http.Request, _
-		renderJSONError(w, err, http.StatusInternalServerError)
-		return
+	if extensionName == "" {
+		extensions, err := h.k8sAPI.GetAllNamespacesWithExtensionLabel(ctx)
```
It looks like when we get this list of extensions from the k8s API, we send all the API info to the frontend. 
![image](https://user-images.githubusercontent.com/2197104/123662391-08d6f380-d7fb-11eb-8aad-e647991a5605.png)

The only info we need in the frontend is the extension name and the namespace. I would extract that from what we receive from the k8s API and send back a list of objects that contain the extension's name and its namespace.

dadjeibaah commented on web/app/js/components/ServiceMesh.jsx:
```diff
@@ -109,6 +118,15 @@ class ServiceMesh extends React.Component {
+    const { extensions } = this.state;
+    let extensionList = [];
+    if (!_isEmpty(extensions.extensions)) {
+      extensionList = extensions.extensions.map(extension => extension.metadata);
```
If we send back only the info we need from the backend, there shouldn't be a need to go through this extra step of filtering.

sannimichaelse replied:
Thanks for feedbacks, would check them out

dadjeibaah reviewed (commented):
This is looking good @sannimichaelse! Thanks for submitting this PR! I left a few comments for your consideration.

dadjeibaah commented:
@sannimichaelse, checking in to see if you have any questions on this PR.

sannimichaelse commented:
> @sannimichaelse, checking in to see if you have any questions on this PR.

Hi @dadjeibaah, no questions really - been busy with final exams - rounding up this week. Should get to it this week. Sorry for any inconvenience

mateiidavid commented on web/srv/api_handlers.go:
```diff
@@ -434,11 +434,44 @@ func (h *handler) handleAPIResourceDefinition(w http.ResponseWriter, req *http.R
+			return
+		}
+
+		type Extension struct {
```
I think we're missing the `namespace` here? looking at issue, it seems the namespace should be present for the extension. Is this something that we want to display?

cc. @dadjeibaah

mateiidavid commented on web/srv/api_handlers.go:
```diff
@@ -434,11 +434,44 @@ func (h *handler) handleAPIResourceDefinition(w http.ResponseWriter, req *http.R
+		extensionObject := Extension{}
+		extensionList := make([]Extension, 0, len(extensions))
+		if len(extensions) > 0 {
+			for _, extension := range extensions {
```
this is a bit nitpicky but I think this block can be a bit confusing w.r.t to the data that's being copied. wdyt about renaming some of these variables to make it a bit clearer to understand what's going on?

for example (feel free to swap with your own)
```suggestion
			for _, installedExtension := range installedExtensions {
			  extension.Name = installedExtension.ObjectMeta.Name
```

It's pretty nitpicky so TIOLI :)

mateiidavid commented on web/srv/api_handlers.go:
```diff
@@ -434,11 +434,44 @@ func (h *handler) handleAPIResourceDefinition(w http.ResponseWriter, req *http.R
+
+		type Extension struct {
+			Name string `json:"name"`
+			UID  string `json:"uid"`
```
What is the purpose of the `UID` here?

sannimichaelse replied:
On the react side, the table needs a unique key while iterating through the values - that's why I decided to add the UID instead of generating a new id since its part of the info

dadjeibaah reviewed (commented):
This is looking great @sannimichaelse. I've left a bunch of comments for you to consider.

dadjeibaah commented:
@sannimichaelse, it looks like you'd also need to rebase your PR against the current main branch since that has some fixes that'll make the integration tests pass.

dadjeibaah reviewed (approved):
Thanks for making those updates. The changes look great and work well. Things look good on my side but It'd be good to get another pair of eyes on the PR as well. 📦 🚢

mateiidavid commented on web/app/js/components/ServiceMesh.jsx:
```diff
@@ -34,6 +34,13 @@ const styles = {
+  {
+    title: <Trans>columnTitleName</Trans>,
+    dataIndex: 'name',
+  },
```
Can we please also show the namespace here as previously discussed?

mateiidavid reviewed (changes requested):
@sannimichaelse this looks really good! The code's looking nice, however, we're still missing the namespace in the table :( would you mind addressing that? after it should be good to merge.

<img width="737" alt="image" src="https://user-images.githubusercontent.com/18265670/129373781-5a29d51a-8d42-4765-8fec-2f9a7d97e6f9.png">

mateiidavid reviewed (approved):
Thanks @sannimichaelse for your effort! PR looks good :)

sannimichaelse commented:
I agree to the DCO for all the commits in this PR
